### ISSUE TITLE
SecureActionButton_OnClick "isKeyPress" parameter is not being passed for SecureActionButtonTemplate, breaking custom action buttons in various ways

### ISSUE THREAD

**Issue by Nevcairiel** [Stanzilla/WoWUIBugs#307]
In 10.0.0.46047 the `isKeyPress` argument started to be used in `SecureActionButton_OnClick` to gate the `useOnKeyDown` condition.

Unfortunately, `isKeyPress` is not being communicated to `SecureActionButton_OnClick` when using a button created from SecureActionButtonTemplate and keybinding it with a `CLICK` type binding, breaking a variety of action button behaviors relying on the down event, including **cast-on-down**, **hold-to-cast**, and **empowered casts**.

In general, it seems like OnClick script handlers have not actually gained this parameter, but its exclusive to `SecureActionButton_OnClick`, which does not work for any custom SecureActionButtons, and only for the FrameXML action buttons which specifically pass this parameter by their virtue of calling this function securely from their keybinding handler.

Impact: Critical. Breaking key functionality for all action bar addons, including the new Evoker empowered spells.

Simple Example:
```lua
local button = CreateFrame("Button", "ClickKeyPressTestButton", UIParent, "SecureActionButtonTemplate")
-- hooking OnClick to get logging of parameters
button:HookScript("OnClick", function(self, button, down, isKeyPress) print(isKeyPress) end)
SetOverrideBindingClick(button, true, "1", "ClickKeyPressTestButton", "LeftButton")
```

Run the above code, try to press 1. If a key is pressed, isKeyPress should be true, otherwise `SecureActionButton_OnClick` fundamentally cannot work right.

Suggested fixes:
- Actually extend the OnClick script handler to pass this argument for all buttons
- Rework `SecureActionButton_OnClick` to not use it (see below)

Although fixing this parameter (as in suggestion 1) would resolve the most immediate issues, please also keep in mind that people have been used to cast-on-down functionality when clicking on buttons as well, and there is no fundamental reason why empower or even hold-to-cast shouldn't work while clicking on a button and holding it down. This might even be an accessibility concern for some, so not relying on a differentation between keybinds and clicks might be best.

In general, these Secure Templates are the cornerstone of many addon functionalities. We can only hope that they remain fully functional for addons to use, not impacted by UI settings (cvars), or using special parameters only accessible by the Default UI.

**Comment by Nevcairiel** [comment]
This issue is now also present on the 10.0.2 beta client.

**Comment by Nevcairiel** [comment]
We've been told part of this is because they don't want hold-to-cast with the mouse, so discussing this and possible solutions, this initially splits into two issues:

(1) Mouse clicks can't do empowered spells (properly), or use on-down casting (and hold-to-cast, but lets say thats intentional)
(2) Addon Keybinds are treated like mouse clicks, with the same limitations as above

One way to fix (2) would be to actually extend OnClick handler to pass that IsKeyPress parameter when its triggered through eg. a CLICK binding (SetBindingClick, etc), and not only have ActionButton code pass it through privileged means that addons cannot do.

This would not fix (1), and for that `SecureActionButton_OnClick` would need to get refined somewhat to only block what is really in need of blocking, instead of blocking all handling of the down event. Ideally this would mean that the C UseAction call (which gets passed the `isKeyPress` parameter) would just not perform hold-to-cast, instead of trying to block it in Lua.

From my POV, if addons couldn't do hold-to-cast in the interrim, that would also be an acceptable compromise, as its new and somewhat niche (eg. only fixing (1)). Long-term of course it would be nice to provide parity.

TL;DR:

Either fixing (1) or (2) would provide short-term relief for addons. Fixing (2) would restore parity with the default UI.

**Comment by Nevcairiel** [comment]
In 10.0.0.46270 this was mostly resolved. Hold-to-Cast is not available to addons right now, but down handling is working again, and hopefully so is hold-release empowered casting (testing pending until this arrives on beta).

I'll leave it open until empowered can be confirmed.

**Comment by mbattersby** [comment]
Is this really resolved? 10.0.0.46340 (Retail) will only trigger actions from SecureActionButtonTemplate where `isDown == GetCVarBool("ActionButtonUseKeyDown")`.

This means addons have no choice but to RegisterForClicks("AnyDown", "AnyUp") and manually abort any PreClick or PostClick handlers if `isDown ~= GetCVarBool("ActionButtonUseKeyDown")`. That is wildly non-intuitive and not a resonable behavior.

In addition the changes have broken /click (`SecureCmdList["CLICK"}` in ChatFrame.lua).

With no button arg, always sets isDown = false
```
/click Button
```
With any button arg, always sets isDown = true
```
/click Button arg
```
which means you have to know in advance the setting of `ActionButtonUseKeyDown` in order to write a macro or use /click.

**Comment by Nevcairiel** [comment]
This bug is solved, there are others affecting this.

**Comment by Nevcairiel** [comment]
Confirmed that Empowered casts also work on the beta. All further issues related to this should track #317
